File: reports/20260826-143732/results/storage/data-plane/python/crud/storage-dp-python-encrypted-uploader/python-azure-tools/baseline/generated-code/key_management.py
"""Envelope-key management backed by Azure Key Vault Keys."""

from __future__ import annotations

import os
from dataclasses import dataclass

from azure.core.credentials import TokenCredential
from azure.core.exceptions import AzureError
from azure.keyvault.keys import KeyClient
from azure.keyvault.keys.aio import KeyClient as AsyncKeyClient
from azure.keyvault.keys.crypto import (
    CryptographyClient,
    KeyWrapAlgorithm,
)
from azure.keyvault.keys.crypto.aio import CryptographyClient as AsyncCryptographyClient
from azure.core.credentials_async import AsyncTokenCredential

AES_KEY_BYTES = 32
WRAP_ALGORITHM = KeyWrapAlgorithm.rsa_oaep_256


class KeyManagementError(RuntimeError):
    """Raised when Key Vault cannot protect or recover a data key."""


@dataclass(frozen=True)
class ProtectedDataKey:
    key_id: str
    algorithm: str
    wrapped_key: bytes


def generate_data_key() -> bytearray:
    """Generate an ephemeral AES-256 key that callers must erase after use."""
    return bytearray(os.urandom(AES_KEY_BYTES))


def erase_data_key(data_key: bytearray) -> None:
    for index in range(len(data_key)):
        data_key[index] = 0


class KeyVaultKeyManager:
    def __init__(
        self,
        key_client: KeyClient,
        credential: TokenCredential,
        key_name: str,
    ) -> None:
        self._key_client = key_client
        self._credential = credential
        self._key_name = key_name

    def protect_data_key(self, data_key: bytes | bytearray) -> ProtectedDataKey:
        try:
            vault_key = self._key_client.get_key(self._key_name)
            if not vault_key.id:
                raise KeyManagementError("Key Vault returned a key without an ID")
            crypto_client = CryptographyClient(vault_key.id, self._credential)
            result = crypto_client.wrap_key(WRAP_ALGORITHM, bytes(data_key))
            return ProtectedDataKey(
                key_id=vault_key.id,
                algorithm=WRAP_ALGORITHM.value,
                wrapped_key=result.encrypted_key,
            )
        except KeyManagementError:
            raise
        except AzureError as exc:
            raise KeyManagementError(
                f"Key Vault could not wrap the data key: {exc}"
            ) from exc

    def recover_data_key(self, protected_key: ProtectedDataKey) -> bytearray:
        try:
            algorithm = KeyWrapAlgorithm(protected_key.algorithm)
            crypto_client = CryptographyClient(
                protected_key.key_id, self._credential
            )
            result = crypto_client.unwrap_key(algorithm, protected_key.wrapped_key)
            if len(result.key) != AES_KEY_BYTES:
                raise KeyManagementError(
                    f"Key Vault returned an invalid {len(result.key)}-byte data key"
                )
            return bytearray(result.key)
        except (ValueError, KeyManagementError):
            raise
        except AzureError as exc:
            raise KeyManagementError(
                f"Key Vault could not unwrap the data key: {exc}"
            ) from exc


class AsyncKeyVaultKeyManager:
    def __init__(
        self,
        key_client: AsyncKeyClient,
        credential: AsyncTokenCredential,
        key_name: str,
    ) -> None:
        self._key_client = key_client
        self._credential = credential
        self._key_name = key_name

    async def protect_data_key(
        self, data_key: bytes | bytearray
    ) -> ProtectedDataKey:
        try:
            vault_key = await self._key_client.get_key(self._key_name)
            if not vault_key.id:
                raise KeyManagementError("Key Vault returned a key without an ID")
            crypto_client = AsyncCryptographyClient(
                vault_key.id, self._credential
            )
            result = await crypto_client.wrap_key(
                WRAP_ALGORITHM, bytes(data_key)
            )
            return ProtectedDataKey(
                key_id=vault_key.id,
                algorithm=WRAP_ALGORITHM.value,
                wrapped_key=result.encrypted_key,
            )
        except KeyManagementError:
            raise
        except AzureError as exc:
            raise KeyManagementError(
                f"Key Vault could not wrap the data key: {exc}"
            ) from exc

    async def recover_data_key(
        self, protected_key: ProtectedDataKey
    ) -> bytearray:
        try:
            algorithm = KeyWrapAlgorithm(protected_key.algorithm)
            crypto_client = AsyncCryptographyClient(
                protected_key.key_id, self._credential
            )
            result = await crypto_client.unwrap_key(
                algorithm, protected_key.wrapped_key
            )
            if len(result.key) != AES_KEY_BYTES:
                raise KeyManagementError(
                    f"Key Vault returned an invalid {len(result.key)}-byte data key"
                )
            return bytearray(result.key)
        except (ValueError, KeyManagementError):
            raise
        except AzureError as exc:
            raise KeyManagementError(
                f"Key Vault could not unwrap the data key: {exc}"
            ) from exc

File: rerun-reports/04-js-ts/20260829-141457/results/storage/data-plane/js-ts/crud/storage-dp-js-ts-encrypted-uploader/js-ts-azure-skills/azure-skill-mcp-microsoft-skill/generated-code/src/keyManagement.ts
import { randomBytes } from "node:crypto";
import type { TokenCredential } from "@azure/core-auth";
import {
  CryptographyClient,
  type KeyClient,
} from "@azure/keyvault-keys";
import {
  describeAzureFailure,
  EncryptedBlobError,
} from "./errors.js";

const DATA_KEY_LENGTH_BYTES = 32;
export const KEY_WRAP_ALGORITHM = "RSA-OAEP-256" as const;

export interface ProtectedDataKey {
  readonly keyId: string;
  readonly wrappedKey: Buffer;
  readonly wrapAlgorithm: typeof KEY_WRAP_ALGORITHM;
}

export class KeyVaultKeyManager {
  public constructor(
    private readonly keyClient: KeyClient,
    private readonly credential: TokenCredential,
    private readonly keyName: string,
  ) {}

  public generateDataKey(): Buffer {
    return randomBytes(DATA_KEY_LENGTH_BYTES);
  }

  public async protectDataKey(dataKey: Buffer): Promise<ProtectedDataKey> {
    if (dataKey.length !== DATA_KEY_LENGTH_BYTES) {
      throw new EncryptedBlobError(
        "cryptography",
        "wrap data key",
        `The data encryption key must be ${DATA_KEY_LENGTH_BYTES} bytes.`,
      );
    }

    try {
      const key = await this.keyClient.getKey(this.keyName);
      if (!key.id) {
        throw new Error("Key Vault returned a key without a versioned key ID.");
      }

      const cryptographyClient = new CryptographyClient(
        key.id,
        this.credential,
      );
      const result = await cryptographyClient.wrapKey(
        KEY_WRAP_ALGORITHM,
        dataKey,
      );

      return {
        keyId: key.id,
        wrappedKey: Buffer.from(result.result),
        wrapAlgorithm: KEY_WRAP_ALGORITHM,
      };
    } catch (error) {
      throw new EncryptedBlobError(
        "key-vault",
        "wrap data key",
        `Azure Key Vault could not protect the data key: ${describeAzureFailure(error)}`,
        { cause: error },
      );
    }
  }

  public async recoverDataKey(
    keyId: string,
    wrappedKey: Buffer,
    wrapAlgorithm: string,
  ): Promise<Buffer> {
    if (wrapAlgorithm !== KEY_WRAP_ALGORITHM) {
      throw new EncryptedBlobError(
        "cryptography",
        "unwrap data key",
        `Unsupported key wrap algorithm: ${wrapAlgorithm}.`,
      );
    }

    this.validateKeyId(keyId);

    try {
      const cryptographyClient = new CryptographyClient(
        keyId,
        this.credential,
      );
      const result = await cryptographyClient.unwrapKey(
        KEY_WRAP_ALGORITHM,
        wrappedKey,
      );
      const dataKey = Buffer.from(result.result);

      if (dataKey.length !== DATA_KEY_LENGTH_BYTES) {
        dataKey.fill(0);
        throw new Error(
          `Key Vault returned an invalid ${dataKey.length}-byte data key.`,
        );
      }

      return dataKey;
    } catch (error) {
      throw new EncryptedBlobError(
        "key-vault",
        "unwrap data key",
        `Azure Key Vault could not recover the data key: ${describeAzureFailure(error)}`,
        { cause: error },
      );
    }
  }

  private validateKeyId(keyId: string): void {
    let parsed: URL;

    try {
      parsed = new URL(keyId);
    } catch (error) {
      throw new EncryptedBlobError(
        "cryptography",
        "validate key ID",
        "Blob metadata contains an invalid Key Vault key ID.",
        { cause: error },
      );
    }

    if (
      parsed.protocol !== "https:" ||
      !/^\/keys\/[^/]+\/[^/]+$/.test(parsed.pathname)
    ) {
      throw new EncryptedBlobError(
        "cryptography",
        "validate key ID",
        "Blob metadata must reference a versioned HTTPS Key Vault key ID.",
      );
    }
  }
}
